### DISCUSSION TITLE
Random playback repeats tracks frequently

### DISCUSSION THREAD

**Discussion by nramsbottom** [MoleMan1024/audiowagon#121] [General]
I have a library of about 1500 tracks and I typically just have it set to play on shuffle. I notice that the same tracks seem to keep playing very frequently, even during fairly short drives (20 minutes or so.)

I've pulled the source and see that the "shuffle" function is capped to 500 random tracks and the randomisation is delegated to the database ```random()``` command.

Is there anything preventing (such as AAOS limitations) a pre-calculated, distinct list of 500 tracks from being generated instead of relying on the database engine?

**Comment by MoleMan1024** [comment]
It's not a limitation of Android but I have limited it to 500 items because the latency just increases more and more with larger lists. And as explained in the FAQs I don't see the point of having huge lists, 500 tracks with approx 3 minutes each is 25 hours, long enough for anyone on a single drive.

I am not aware that the sqlite `random()` function has any issues. Not sure what you mean with "pre-calculating" a "random" list, but it seems against the spirit of having a "random" function.

Then again it is random and it might just be coincidence. But I can have a look and do some test runs maybe if I see a similar behaviour.

**Comment by MoleMan1024** [comment]
Here is a plot of some tests (on Windows) for the sqlite random() function for a database of 1556 tracks (selecting 500 tracks at random iterating 200 times accessed via "shuffle all" media item in media browser track view). It shows a (skewed) normal distribution of the similarity of the resulting playback queues, as I would expect:

![image](https://github.com/MoleMan1024/audiowagon/assets/70017071/a87a2e4d-ad36-471a-88f7-983f0f23b1f8)

Interestingly this is faster but generates more similar lists:
`SELECT * FROM track WHERE trackId IN (SELECT trackId FROM track ORDER BY RANDOM() LIMIT 500)`

![image](https://github.com/MoleMan1024/audiowagon/assets/70017071/971f9570-a816-464d-921e-10f0284d6225)

**Comment by MoleMan1024** [comment]
This one provides better speed and has less similar results, I might switch to this:

`SELECT * FROM track as t1 JOIN (SELECT trackId FROM track ORDER BY RANDOM() LIMIT 500) as t2 ON t1.trackId = t2.trackId`

![image](https://github.com/MoleMan1024/audiowagon/assets/70017071/2274285c-0ce8-4e87-8478-b6eb3050c19a)

**Comment by nramsbottom** [comment]
What I was thinking with a "pre-generated" random play list is something like a pseudo-random generated list of distinct track IDs. It would be computationally expensive to create, but once created it could be stored or regenerated.

I would have a crack at implementing this myself, but my Java is... very bad.
